### PR TITLE
Graphics: Adapt aspect ratio when SBS/TAB 3D is used

### PR DESCRIPTION
**Pull request by BryanJacobs**
I use Dolphin to play games in a mix of side-by-side 3D and flat 2D. I make use of the hotkey functionality to disable/enable 3D. My 3D glasses have two 1080p screens in them, so when in SBS mode, the aspect ratio of the viewport is 32:9 (each eye is 16:9).

One problem I've noticed is that content being displayed at 16:9 in 2D suddenly becomes squared off when 3D is enabled. This is surprising; if it was widescreen content, why wouldn't it remain widescreen?

In my opinion the "aspect ratio" setting under Graphics is intended to be the aspect ratio of the CONTENT, not the viewport. So when enabling side-by-side 3D, the aspect ratio of the viewport should be double the given value. And when enabling top-and-bottom 3D, the aspect ratio of the viewport should be half the given value.

This change makes that happen; if content was being displayed as 16:9, toggling on SBS 3D will make it 32:9, or 16:9 per eye.

This could be done with a separate config option, but I think it's likely what 3D users would want... if someone wants to force half-SBS, they could do that by fixing an aspect ratio of 8:9.

### CODE REVIEW

**Comment by MayImilae** [comment]
So I can't particularly review this, I don't have the hardware to test it and analogue aspect ratios are a nightmare I don't want to walk into. However, I'm pretty certain that leaving the aspect ratio alone was intentional. By doing so, Dolphin supports "Half Side by Side" (Half SBS) which is (or at least was) more common than Full SBS. I believe going with Half SBS was a deliberate choice. @CrossVR answer this if you can.

I don't know how many of our users are using Half SBS or how many would want Full SBS, or which is currently more common for 3D displays. However, rather than remove Half SBS for Full SBS, I think it would be best to just include both Half SBS and Full SBS modes in the Stereoscopic 3D Mode selection. I personally would be much more comfortable with that.

**Comment by BryanJacobs** [comment]
I think you could review this without the hardware, because SBS and TAB are literally just, uh, two images side by side or above and below each other. You can see if it's right on a 2D monitor.

I can rework this to have a toggle box for "full SBS" that enables this behavior, under 3D settings. That's a slightly bigger/more intrusive change - is that what you'd like?

**Comment by MayImilae** [comment]
UI and UX design is something I am very qualified to comment on!

My preference would be to have the "Stereoscopic 3D Mode" combo's "Side by Side" entry replaced with "Half Side by Side" and "Full Side by Side" entries. This is clean and simple.

Alternatively you could have a radio for Half SBS and Full SBS that is only added to the Stereoscopy section when SBS is enabled. When other Stereoscopy modes are selected, SBS switching would just not appear. This is a little more complex to implement but it's been done elsewhere in Dolphin's Qt UI and it does the job.

Basically, I'd really like to avoid the scenario where we have some stereoscopy options that are global for all stereoscopy settings, and some that are disabled until a certain mode is used. It's gross.

**Comment by BryanJacobs** [comment]
> UI and UX design is something I am very qualified to comment on!
> 
> My preference would be to have the "Stereoscopic 3D Mode" combo's "Side by Side" entry replaced with "Half Side by Side" and "Full Side by Side" entries. This is clean and simple.

Do you also have a preference for what happens to the Settings > Hotkeys > 3D page, which currently has "toggle 3D SBS" and "toggle 3D TAB"?

> Alternatively you could have a radio for Half SBS and Full SBS that is only added to the Stereoscopy section when SBS is enabled. When other Stereoscopy modes are selected, SBS switching would just not appear. This is a little more complex to implement but it's been done elsewhere in Dolphin's Qt UI and it does the job.

This sounds better for the user to me.

> 
> Basically, I'd really like to avoid the scenario where we have some stereoscopy options that are global for all stereoscopy settings, and some that are disabled until a certain mode is used. It's gross.

Yes, I can hide a checkbox for "apply full resolution to each eye" where irrelevant. WDYT?

**Comment by MayImilae** [comment]
> Yes, I can hide a checkbox for "apply full resolution to each eye" where irrelevant. WDYT?

That sounds fine to me.

**Comment by CrossVR** [comment]
Just to add some context here, the 3D stereoscopy feature was developed during a time where 3DTVs were still common. These TVs didn't actually support a larger display mode for 3D, but simply cut the picture in half and then stretched it across the entire screen.

This didn't affect the quality though, because the polarization layer on top of the screen would send half the pixels to each eye thereby halving the effective resolution anyway. What was important back then was that you selected the right mode based on whether the polarization layer on your TV split each column of pixels (SBS) or each row of pixels (TAB) between the eyes.

**Comment by BryanJacobs** [comment]
I've added a configuration radio button for "per eye resolution" with two choices - "Full" and "Half". When Half is chosen, Dolphin behaves as today. When Full is chosen, Dolphin doubles (for SBS) or halves (for TAB) the aspect ratio of the presented content.

This option is entirely hidden (not just disabled) if a selection other than TAB or SBS is made in the Enhancements editor.

The setting is persisted in the GFX.ini file and can also be saved per-game.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 1]
```diff
@@ -589,6 +611,10 @@ void EnhancementsWidget::AddDescriptions()
   static const char TR_3D_SWAP_EYES_DESCRIPTION[] = QT_TR_NOOP(
       "Swaps the left and right eye. Most useful in side-by-side stereoscopy "
       "mode.<br><br><dolphin_emphasis>If unsure, leave this unchecked.</dolphin_emphasis>");
+  static const char TR_3D_PER_EYE_RESOLUTION_DESCRIPTION[] =
+      QT_TR_NOOP("Whether each eye gets full or half image resolution when using side-by-side "
+                 "or above-and-below 3D."
+                 "<br><br><dolphin_emphasis>If unsure, leave set to full.</dolphin_emphasis>");
```
Comment (JosJuice) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:617]: Just for consistency with other descriptions:

```suggestion
                 "<br><br><dolphin_emphasis>If unsure, select Full.</dolphin_emphasis>");
```

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp [hunk 2]
```diff
@@ -242,8 +258,14 @@ void EnhancementsWidget::ConnectWidgets()
     m_block_save = true;
     m_configure_color_correction->setEnabled(g_Config.backend_info.bSupportsPostProcessing);
     LoadPPShaders();
-    m_block_save = false;
 
+    auto current_stereo_mode = Config::GetUncached(Config::GFX_STEREO_MODE);
```
Comment (JosJuice) [inline Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp:262]: Is there a reason why you can't use the regular `Config::Get`? `Config::GetUncached` should only really be needed internally in `Config::Get`.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 1 above)
Comment (CrossVR) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: Also, given that Half has been the default for years, the "if unsure" recommendation should probably be Half.

Comment (BryanJacobs) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: > Also, given that Half has been the default for years, the "if unsure" recommendation should probably be Half.

I disagree. I'm willing to change this if you feel strongly about it, but...

Today, if you buy a device that supports side-by-side 3D, it likely does so with two identical screens. So far as I know there aren't more "half SBS" products being produced. If you select "Half" with one of those devices, what you get is something that _sort of_ looks right - it's 3D... but you're not only losing half the resolution in the visible content area, you've also got letterboxing meaning the content doesn't occupy your viewport properly.

I think we should recommend to those who are unsure that they select Full, and if someone knows they have a half-SBS device (and how could they not know?!), they'll know well enough to ignore the default.

Comment (iwubcode) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: I'm indifferent to the default.  My thought is that people that partake in stereo 3d are usually quite tech savy.  If they see the wrong 3d output, they will likely understand they need to mess with the settings.  Additionally, given most of the new tech devices that offer 3d (Lume Pad, Looking Glass, some of the special monitors with eye tracking) don't work with Dolphin, most of the tech being used is going to be older regardless of which default we choose.

**Comment by iwubcode** [comment]
Assuming this change is complete.  Please squash your commits

**Comment by BryanJacobs** [comment]
> Assuming this change is complete. Please squash your commits

Done.

File: Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp (hunk 1 above)
Comment (iwubcode) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: I will say, that this should probably match the default.  Both the default of the enum and the default in the settings is 'Half'.  I didn't do a full look but it would seem odd to have a different default than what is recommended.

Comment (BryanJacobs) [reply Source/Core/DolphinQt/Config/Graphics/EnhancementsWidget.cpp]: This was exactly what I was trying to say. The current Dolphin behaviour is "Half". But I think it's better to use Full, because using Full is more commonly the correct setting (and as time goes on, Half will be the right setting less and less often).

I left the config defaulted to Half to preserve the existing behaviour, but recommended the use of Full for that reason.

I don't particularly mind if we change either the default setting (breaking backwards compat) or the recommendation (recommending users do something that will be less likely the right thing as the years go by). Whatever you like.

**Review by mbc07 — CHANGES_REQUESTED** [review]
The functionality change is fine, but I strongly disagree with the UI choice here.

Having the half/full resolution choice as radio boxes inside a group box which is already inside a parent group box is clunky, unnecessarily takes more space than what it's actually needed and (IMHO) is just ugly. Localization-wise, I can also picture long translations making the group box grow even larger and awkwardly positioned than it already is in English.

With that said, please convert the "Per Eye Resolution" group box and its content into a simple checkbox that's hidden until "side by side" or "top and bottom" is selected in the stereoscopy mode dropdown (there's UI code for that in the GameCube config tab from Dolphin's general settings you can use as reference). 

Regarding the behavior, I think checkbox enabled = full resolution, checkbox disabled = half resolution should work, but please put it below the existing "Swap Eyes" checkbox, so the showing/hiding behavior doesn't unnecessarily move settings visible all times.

 Something like this mockup (feel free to change the wording if you want): 
image

**Comment by BryanJacobs** [comment]
> With that said, please convert the "Per Eye Resolution" group box and its content into a simple checkbox that's hidden until "side by side" or "top and bottom" is selected in the stereoscopy mode dropdown. Regarding the behavior, I think checkbox enabled = full resolution, checkbox disabled = half resolution should work, but please put it below the existing "Swap Eyes" checkbox, so the showing/hiding behavior doesn't unnecessarily move settings visible all times.

I've done exactly as you asked, and rebased/squashed atop the latest master. I've also changed the documentation to say "if unsure, leave this unchecked" to match the backwards-compatible default.

**Comment by mbc07** [comment]
@dolphin-emu-bot rebuild

**Review by mbc07 — APPROVED** [review]

**Review by iwubcode — APPROVED** [review]

**Review by MayImilae — APPROVED** [review]
UI LGTM. Untested.
